sdk/eventhub/Azure.Messaging.EventHubs/src/Core/TransportProducerPool.cs

```diff
@@ -0,0 +1,263 @@
+                                        TimeSpan.FromMinutes(PerformExpirationPeriodAsMinutes),
+                                        TimeSpan.FromMinutes(PerformExpirationPeriodAsMinutes));
+        }
+
+        /// <summary>
+        ///   It returns a <see cref="PooledProducer"/> corresponding to a given partition
+        ///   and it extends its sliding expiration.
+        /// </summary>
+        ///
+        /// <param name="partitionId">The zero-based unique identifier of a partition associated with the Event Hub.</param>
+        /// <param name="connection">The <see cref="EventHubConnection"/> to use as the basis for delegation of client-type operations.</param>
+        /// <param name="retryPolicy">The <see cref="EventHubsRetryPolicy"/> to use as the basis for retrieving the information.</param>
+        ///
+        /// <returns>A <see cref="PooledProducer"/> corresponding to the partitionId passed in as parameter.</returns>
+        ///
+        /// <remarks>
+        ///   There is a slight probability that the returned producer may be closed at any time
+        ///   after it is returned and the caller may want to handle that scenario.
+        /// </remarks>
+        ///
+        public PooledProducer GetPartitionProducer(string partitionId,
+                                                   EventHubConnection connection,
+                                                   EventHubsRetryPolicy retryPolicy)
+        {
+            Argument.AssertNotNull(partitionId, nameof(partitionId));
+            Argument.AssertNotNull(connection, nameof(connection));
+            Argument.AssertNotNull(retryPolicy, nameof(retryPolicy));
+
+            var identifier = Guid.NewGuid().ToString();
+
+            var item = Pool.GetOrAdd(partitionId, id => new PoolItem(connection.CreateTransportProducer(id, retryPolicy)));
+
+            // A race condition at this point may end with an CloseAsync called on
+            // the returned PoolItem if it had expired. The probability is very low and
+            // possible exceptions should be handled by the invoking methods.
+
+            if (!item.ActiveInstances.TryAdd(identifier, 0) || item.Producer.IsClosed)
+            {
+                identifier = Guid.NewGuid().ToString();
+                item = Pool.GetOrAdd(partitionId, id => new PoolItem(connection.CreateTransportProducer(id, retryPolicy)));
+                item.ActiveInstances.TryAdd(identifier, 0);
+            }
+
+            item.RemoveAfter = DateTimeOffset.UtcNow.Add(PoolItem.DefaultRemoveAfterDuration);
+
+            return new PooledProducer(item.Producer, cleanUp: producer =>
+            {
+                if (Pool.TryGetValue(partitionId, out PoolItem pooledItem))
+                {
+                    pooledItem.RemoveAfter = DateTimeOffset.UtcNow.Add(PoolItem.DefaultRemoveAfterDuration);
+                }
+
+                if (!item.ActiveInstances.TryRemove(identifier, out _))
```

Comment:
I don't know that we need the `if` statement here.   Just calling `TryRemove` and decorating with the comment should be sufficient and won't rely on compiler optimizations to recognize and remove the empty branch conditional.